[WiP] Initial protos for mutation filter

gbrail opened the pull request:
This introduces a new filter called the "external mutation filter."
It is intended to allow an external service to be able to operate as
if it were part of the filter chain using a gRPC stream. It is intended
to support a variety of use cases in which processing of HTTP requests
and responses by an external service is desired.

We'll follow up with a public design proposal when others have had
more of an opportunity to look at the proposed protobufs.

Part of #5209

Signed-off-by: Gregory Brail <gregbrail@google.com>

htuch commented:
Relates to https://github.com/envoyproxy/envoy/issues/5209. @gbrail I'm going to tag this as [WiP] in title, since we'll be using this mostly for discussion and need to have the public design document as next step. Thanks.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,203 @@
+  bool end_of_stream = 2;
+}
+
+// This message completely replaces the request or response.
```
Maybe CompleteRequestResponse?

gbrail replied:
Yeah, let me look at the name.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,203 @@
+    Body body = 3;
+
+    // The HTTP trailers, for either the request or response path, depending
+    // on where we are in the filter lifecycle. This is only sent if the
```
The order of downstream/upstream trailers isn't fixed, an upstream might send trailers before downstream or vice versa.

gbrail replied:
OK -- I'll have to think about how to prevent the asynchronous nature of all this from getting out of control for the server implementer.

gbrail replied:
I assume it's also possible that encodeHeaders could come before decodeData, if the upstream has returned an error before the downstream has sent the whole body, right?

htuch replied:
Right.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,203 @@
+    // Information about the HTTP request, sent after the HTTP headers have been
+    // received on the "encoding" path, but before the body is available.
+    // This is the same data sent to an "ext_authz" service.
+    auth.v3.AttributeContext request = 1;
```
Not sure I fully follow why this is only available when the upstream is repsonding and we're encoding.. why can't we have request metadata or attributes earlier?

gbrail replied:
I was just trying to explain that this message is sent when decodeHeaders is called on the filter. I'll see about changing the comment.

htuch replied:
I think in general referencing Envoy filter lifecycle events makes sense in this file, rather than more abstract concepts of request/response, since for streaming some of these break down.

htuch reviewed (changes requested):
Flushing some comments, thanks for driving this super important feature!
Tagging @PiotrSikora as well for thoughts.
/wait

gbrail commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,203 @@
+  // by the client is a "*BODY" request.
+  BodyMutation body_mutation = 5;
+
+  // Add new metadata to the Envoy context
```
TBH I copied this directly from the ext_authz filter. It has a specifier in its configuration and there's a specifier in the configuration of this filter as well. But I haven't dived in to see what it all means yet.

htuch replied:
Yeah, I reckon adding additional comments (and even backfilling them to ext_authz) makes sense. We want this to be as standalone and easy to reason about as possible. The protobuf comments you see here get turned into Envoy user facing docs via tooling, see https://www.envoyproxy.io/docs/envoy/latest/api-v3/api.

gbrail commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,203 @@
+  google.protobuf.Struct new_metadata = 6;
+}
+
+// This message is sent downstream to the server when the HTTP response has been
```
Just encodeHeaders.

mpuncel commented on api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,80 @@
+
+  // If true, then do nothing if the remote service RPC fails.
+  // Otherwise, return a 500 error if the gRPC call fails.
+  bool failure_mode_allow = 3;
```
this would also call into question the name of the filter, but would it make sense to support a "read only" mode? then this filter could be used for logging and not allowed to actually mutate anything? This could also be implemented with `failure_mode_allow` being true and the service always returning the `CONTINUE_*` responses, but it might be nice to have a guarantee against the upstream modifying anything in certain cases?

htuch replied:
@gbrail I missed the most recent updates, is this now ready for a full review again? I can take a look tomorrow.

I think we really should change the name. As @mpuncel points out, it won't always mutate. Also, it makes it hard to discover; we had a question at EnvoyCon today about whether there is a "gRPC service for offloading compute" as an alternative to Wasm, I shared this PR. I vote for "gRPC remote filter" (grpc_remote) or "remote filter" (remote), but there are possibly better names.

mattklein123 replied:
+1 commented on this above.

gbrail replied:
"external processing filter," or "ext_proc" is the leading candidate.

gbrail commented:
There are two possible ways to handle read-only mode.

First, a server can consume the gRPC stream and respond without asking
Envoy to change anything. The server could even respond before
actually doing the work to log or whatever if it wanted to reduce latency
-- that's effectively a read-only mode.

Second, the configuration protobuf also includes an "async mode." The idea
there is that Envoy will send messages downstream to the remote server, but
won't wait for the remote server to respond. This will reduce latency even
more (if not eliminate it). However, we'll need some sort of flow control
to prevent running out of memory in Envoy and the specific mechanism for
that is TBD.

On Thu, Oct 15, 2020 at 5:24 PM htuch <notifications@github.com> wrote:

> *@htuch* commented on this pull request.
> ------------------------------
>
> In api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r505935850>:
>
> > +//   headers, bodies, or both, and then decide to switch to a synchronous processing
> +//   mode, either permanently or temporarily.
> +
> +// [#next-free-field: 7]
> +message Mutation {
> +  // Configuration for the gRPC service that the filter will communicate with.
> +  // The filter supports both the "Envoy" and "Google" gRPC clients.
> +  config.core.v3.GrpcService grpc_service = 1;
> +
> +  // The version of the API supported by the gRPC service.
> +  config.core.v3.ApiVersion transport_api_version = 2
> +      [(validate.rules).enum = {defined_only: true}];
> +
> +  // If true, then do nothing if the remote service RPC fails.
> +  // Otherwise, return a 500 error if the gRPC call fails.
> +  bool failure_mode_allow = 3;
>
> @gbrail <https://github.com/gbrail> I missed the most recent updates, is
> this now ready for a full review again? I can take a look tomorrow.
>
> I think we really should change the name. As @mpuncel
> <https://github.com/mpuncel> points out, it won't always mutate. Also, it
> makes it hard to discover; we had a question at EnvoyCon today about
> whether there is a "gRPC service for offloading compute" as an alternative
> to Wasm, I shared this PR. I vote for "gRPC remote filter" (grpc_remote) or
> "remote filter" (remote), but there are possibly better names.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r505935850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I27OXTFQ75CAX4R3T73SK6HFNANCNFSM4R5FQWUA>
> .
>

gbrail commented:
I have time tomorrow to work on this, and I'll see about changing the name.
I do think that either "grpc_remote" or just "remote_filter" would be
appropriate.

On Thu, Oct 15, 2020 at 5:24 PM htuch <notifications@github.com> wrote:

> *@htuch* commented on this pull request.
> ------------------------------
>
> In api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r505935850>:
>
> > +//   headers, bodies, or both, and then decide to switch to a synchronous processing
> +//   mode, either permanently or temporarily.
> +
> +// [#next-free-field: 7]
> +message Mutation {
> +  // Configuration for the gRPC service that the filter will communicate with.
> +  // The filter supports both the "Envoy" and "Google" gRPC clients.
> +  config.core.v3.GrpcService grpc_service = 1;
> +
> +  // The version of the API supported by the gRPC service.
> +  config.core.v3.ApiVersion transport_api_version = 2
> +      [(validate.rules).enum = {defined_only: true}];
> +
> +  // If true, then do nothing if the remote service RPC fails.
> +  // Otherwise, return a 500 error if the gRPC call fails.
> +  bool failure_mode_allow = 3;
>
> @gbrail <https://github.com/gbrail> I missed the most recent updates, is
> this now ready for a full review again? I can take a look tomorrow.
>
> I think we really should change the name. As @mpuncel
> <https://github.com/mpuncel> points out, it won't always mutate. Also, it
> makes it hard to discover; we had a question at EnvoyCon today about
> whether there is a "gRPC service for offloading compute" as an alternative
> to Wasm, I shared this PR. I vote for "gRPC remote filter" (grpc_remote) or
> "remote filter" (remote), but there are possibly better names.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r505935850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I27OXTFQ75CAX4R3T73SK6HFNANCNFSM4R5FQWUA>
> .
>

htuch commented on api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,80 @@
+
+  // For each property name, select a matching property value to attach to the
+  // request message. Property names are the same ones supported in the WebAssembly
+  // runtime.
```
This needs to link to some documentation. Most users aren't going to be able to understand what these are.

mattklein123 replied:
+1, also, I don't think we should use a string here. At least do a wrapper message with a string name field. This will allow us to add additional configuration later if needed. Same for response properties.

htuch commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+    // does not wait for a response. By default the request and response headers,
+    // all body chunks, and trailers are sent. An AsyncParams object can be used
+    // to customize this behavior.
+    ASYNC = 1;
```
I like the flexibility here and recall the design discussion. I'd like to verify with @mattklein123 that we're not doubling up on what the TAP filter does here; clearly there is some overlap, but here the intention is that we're planning to use this for some async request/response monitoring sink, rather than debug.  @mattklein123 WDYT?

mattklein123 replied:
I think it's fine to separate this out.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server a chunk of the message body.
```
Presumably this happens 1 or more times.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+    // never resume processing the filter chain. In practice, this will always
+    // be "true" unless the server or the user has configured the filter
+    // to run in "async mode."
+    bool response_required = 1;
```
Should this be outside the `oneof`?

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+// before closing the stream.
+// [#next-free-field: 8]
+message MutationResponse {
+  enum Command {
```
We generally prefer in the Envoy API that when there are complex enums, that a `oneof` is used instead, with messages (even if empty) for the individual values. This allows grouping of state with the options in a tagged union and future proofs.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+}
+
+// Change all or part of the message body by replacing the portion
+// of the original body denoted by "offset" and "length" with "body."
```
Is this relative to the start of the body or the start of the last chunk?

mattklein123 replied:
I assume last chunk?

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+message BodyMutation {
+  int32 offset = 1;
+
+  int32 length = 2;
```
Nit: in the Envoy API style, we prefer unsigned when values can't be negative.

htuch commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+// then the change is a temporary change that will be un-done after the specified
+// timeout. (If multiple requests arrive after a timer has been set, then the results
+// may or may not reflect the last one received.)
+message ProcessingModeChange {
```
I think this is one of the clear ways this differs from TAP CC @mattklein123

htuch reviewed (commented):
This is looking fairly baked, would like @envoyproxy/api-shepherds to take a deeper look, this is going to be pretty heavily used when it's introduced.

PiotrSikora commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+  bool request_headers = 1;
+
+  // If true, send request body chunks asynchronously
+  bool request_body = 2;
```
Please add "chunks" in the name, otherwise people will get confused whether this is complete buffered request body or not (yes, even if it's documented in the line above).

PiotrSikora commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+
+  // If true, send response trailers asynchronously
+  bool response_trailers = 6;
+}
```
This is missing request/response `HTTP/2 METADATA` frames.

mattklein123 replied:
+1. Also, do we need to stream out reset if the stream doesn't finish? Probably? How is premature closure indicated?

PiotrSikora commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+  bool response_headers = 4;
+
+  // If true, send response body chunks asynchronously
+  bool response_body = 5;
```
Ditto.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  oneof r {
```
`oneof` seems like an unnecessary limitation. Since this is sent to an external service, which might be a few ms away, it could be useful e.g. to send a complete request (i.e. request headers + complete request body + request trailers), which isn't possible with `oneof`.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+  // The HTTP URL scheme, such as `http` and `https`.
+  string scheme = 6;
+
+  // Information about the upstream service's connection
```
Upstream or downstream?

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+
+    // The server is done with the request. Continue processing
+    // and do not consult the server again for this request.
+    CONTINUE_REQUEST = 1;
```
This doesn't match Envoy's use of `Continue`.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+
+    // The server is done with the response. Do not consult the server
+    // again.
+    CONTINUE_RESPONSE = 2;
```
Ditto.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+
+    // Request that Envoy will buffer the entire request message body in memory
+    // subject to the currently-configured body buffer limit, and send a
+    // single Body message downstream containing the entire message body.
```
What if it runs out of that body buffer limit? The updated Proxy-Wasm ABI will call into request body callbacks again in that case.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+    // single Body message downstream containing the entire message body.
+    // It is still possible for the server to modify the message headers
+    // with the next upstream message.
+    SEND_REQUEST_BODY = 5;
```
`s/SEND_REQUEST_BODY/SEND_BUFFERED_REQUEST_BODY/`

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+    SEND_REQUEST_BODY = 5;
+
+    // Request that Envoy buffer and send the response body in a single chunk.
+    SEND_RESPONSE_BODY = 6;
```
Ditto.

PiotrSikora commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,275 @@
+
+// This message contains the message body that Envoy sends to the mutation service.
+message Body {
+  bytes body = 1;
```
This could probably be a vector of body chunks, otherwise Envoy will have to linearize it before sending to the external service.

mattklein123 replied:
+1

gbrail replied:
I was planning on using the "StopIterationAndBuffer" feature, which causes Envoy to do just this, but I can see the point.

However, if we look above to the "BodyMutation" proto, it changes the message body based on the size of the last chunk that we sent the server. So in that case, either we need a way to separately mutate each chunk (ick) or we tell the user that the BodyMutation is relative to the list of chunks shown here assuming that they are all concatenated together.

So I think this is a matter of making things easier to implement on the server side versus making Envoy as efficient as we can.

PiotrSikora reviewed (commented):
Nice work, thanks! The one issue here is that this is adding quasi-stable API that's similar-yet-different than the Proxy-Wasm ABI, which might be quite confusing to the end-users.

Ideally, they should be able to as pass-thru both ways, i.e. we should able to create an external mutation filter that calls remote service that implements Proxy-Wasm ABI, and we should be able to create Proxy-Wasm extension in Envoy that calls out to an external mutation service, without the need to do any translations.

mattklein123 commented on api/envoy/extensions/filters/http/mutation/v3alpha/BUILD:
```diff
@@ -0,0 +1,12 @@
+# DO NOT EDIT. This file is generated by tools/proto_format/proto_sync.py.
```
+1, was going to suggest a name change. "offload" is probably fine though we should try to think of other names also. "offload_via_api" "offload_via_grpc" etc.

gbrail replied:
I am going going to propose "external processing filter," or "ext_proc" for short. But let's get through the many comments on this PR because when I change the name I think GitHub will toss many of the comments. LMK if that name is not OK!

mattklein123 commented on api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,80 @@
+  config.core.v3.ApiVersion transport_api_version = 2
+      [(validate.rules).enum = {defined_only: true}];
+
+  // If true, then do nothing if the remote service RPC fails.
+  // Otherwise, return a 500 error if the gRPC call fails.
```
The bi-di stream can fail/timeout at any time. We need a few things here:
1) Reframe this field in terms of any time failure.
2) Think about individual stream element timeouts as generic stream idle timeouts may not cut it.

gbrail replied:
I'll add a timeout to the configuration that will affect every request/response between the proxy and the external processor.

mattklein123 commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+
+message ProcessingMode {
+  enum Mode {
+    // In the default "sync" mode, a message is sent to the filter server
+    // as soon as decodeHeaders is called, and the filter waits for a response
+    // from the server before continuing.
+    SYNC = 0;
```
Per my other comment we will need per-op timeouts. It might be fine to have a timeout for each ACK/mutation, or we might end up needing per-op timeouts? Can you flesh that out?

mattklein123 commented on api/envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,63 @@
+  Mode mode = 1;
+
+  // The optional AsyncParams object controls how data is delivered
+  // to the server in ASYNC mode. It is ignored in SYNC mode.
```
Why is it ignored in SYNC mode? Perhaps for performance reasons the user only wants to send over headers and mutate them and then stream everything else? I think it would be more flexible to rename this `Params` or `DataConfiguration` or something and use in both modes?

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+message MutationRequest {
+  // If true, then Envoy requires that the server send one MutationResponse message
+  // back up the stream in response to this message -- otherwise, Envoy will
+  // never resume processing the filter chain. In practice, this will always
+  // be "true" unless the server or the user has configured the filter
+  // to run in "async mode."
+  bool response_required = 1;
```
Would it better to just send the mode to the server? I think the intention would be more clear?

gbrail replied:
I'm trying to avoid making every server author understand a complex set of calculations. The idea is that the proxy sends one message to the server, and there's a flag right there that tells the server either that the message was informational, or that the proxy needs to respond in a timely way (because the server has presumably suspended filter execution while it waits for a response).

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+  // to run in "async mode."
+  bool response_required = 1;
+
+  // Information about the HTTP request, sent when the decodeHeaders callback
+  // is invoked on the Envoy filter.
+  Request request = 2;
+
+  // Information about the HTTP response, sent when the encodeHeaders callback
+  // is invoked on the Envoy filter.
+  Response response = 3;
+
+  // A chunk of the HTTP request body. This is only sent if the server
+  // requested it.
+  Body request_body = 4;
+
+  // A chunk of the HTTP response body, if the server requested it.
+  Body response_body = 5;
+
+  // The HTTP trailers for the request path. These are only sent if
+  // the server requested them.
+  config.core.v3.HeaderMap request_trailers = 6;
+
+  // The HTTP trailers for the response path, sent if the server
+  // requested them.
+  config.core.v3.HeaderMap response_trailers = 7;
```
Should this be inside a oneof for clarity?

gbrail replied:
Funny, because I just changed it from a oneof to a non-oneof.

In my mind each message will include only one of the above, but someone earlier suggested there might be a future version of the protocol in which the proxy can wait until it has all the bits and send them in only one message. For instance, maybe we have a configuration mode that sends the headers and the first body chunk in one message to optimize for a server that always needs to look at both before deciding what to do.

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+message ContinueResponseCommand {
+}
+
+// Request that Envoy next send a Body message downstream
```
```suggestion
// Request that Envoy next send a Body message
```

gbrail replied:
Ack.

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+message SendBufferedResponseBodyCommand {
+}
+
+// Request that Envoy will send the HTTP request trailers when they are
+// available. Once trailers have been requested, no more bodies
+// will be delivered, so this must be sent after all desired body chunks are
+// desired.
```
How are no trailers / end of stream indicated?

gbrail replied:
I'll update the comment to clarify that an empty set of trailers will be sent if trailers are requested but none are present. (And I'll figure out later whether the HTTP filter model is actually capable of doing that.)

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+  // to the client with no additional filter processing.
+  CompleteResponse complete_response = 10;
+
+  // Add or replace HTTP headers. This will be ignored after SendXXXBodyChunkCommand has
+  // been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated config.core.v3.HeaderValueOption set_headers = 11;
+
+  // Remove these HTTP headers. This will be ignored after SendXXXBodyChunk
+  // has been sent. If SendXXXTrailers was previously sent, then
+  // modify the trailers instead.
+  repeated string remove_headers = 12;
+
+  // Replace some or all of the last message body received with a new
+  // set of bytes. This will be ignored unless the previous message sent
+  // by the client is a "Body" request.
+  BodyMutation body_mutation = 13;
```
It's not clear from an API perspective where these apply. I think all of these should be part of relevant commands above?

gbrail replied:
This has gotten confusing (same as for the "oneof" conversation above). This is basically a request / response protocol on top of a gRPC stream. I'm going to try and reorganize this proto to that effect and see it it helps make it easier for an external server developer to serve this protocol properly without breaking the proxy.

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+// first received.
+// [#next-free-field: 9]
+message Request {
+  // A unique ID that will match the one sent in the "Response" message
+  string id = 1;
```
Is this needed given that all of this happens on a single bi-di stream?

gbrail replied:
Nope - I will remove it.

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+  // A unique ID that will match the one sent in the "Response" message
+  string id = 1;
+
+  // The HTTP request method, such as `GET`, `POST`.
+  string method = 2;
```
This is redundant given you are sending the headers below. Same for all the other specific headers. I would remove them all.

gbrail replied:
I copied ext_authz. But I agree and I'm happy to remove some complexity.

mattklein123 commented on api/envoy/service/mutation/v3alpha/mutation.proto:
```diff
@@ -0,0 +1,302 @@
+  // A unique ID that will match the one sent in the "Request" message
+  string id = 1;
+
+  // The HTTP status on the response
+  int32 status_code = 2;
```
redundant

gbrail replied:
Sure, but useful? There are similar redundant fields in the request message that I copied from the ext_authz filter.

mattklein123 reviewed (changes requested):
Very cool! Flushing out some comments from a first pass. I will take another pass once we sort through these! Thank you.

/wait

gbrail commented:
This latest revision refactors the protocol a bit, and is renamed to "external processing filter," or "ext_proc".

I also created a public document describing the filter:

https://docs.google.com/document/d/1IZqm5IUnG9gc2VqwGaN5C2TZAD9_QbsY9Vvy5vr9Zmw/edit#heading=h.4bti0lhcgu9m

Would you all like to keep iterating on this, or should I squash this and create a new PR?

gbrail commented:
@htuch I could use advice on the next step now that I renamed most of the files.

mattklein123 commented:
@gbrail I will dive back into this next week. This is complicated enough though that it might make sense to focus on the doc first? Should we review that first and then come back to this?

gbrail commented:
Sure -- I'm also prototyping to make sure that the protos actually work.

On Fri, Oct 23, 2020 at 5:38 PM Matt Klein <notifications@github.com> wrote:

> @gbrail <https://github.com/gbrail> I will dive back into this next week.
> This is complicated enough though that it might make sense to focus on the
> doc first? Should we review that first and then come back to this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#issuecomment-715644438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I22PT5CYQVXAQBT2C6TSMIOZXANCNFSM4R5FQWUA>
> .
>

htuch commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,89 @@
+
+  // For each property name, select a matching property value to attach to the
+  // request message. Property names are the same ones supported in the WebAssembly
+  // runtime.
```
I think my comments were lost, but I still don't know what these are and we need some RST links here. @PiotrSikora can you help?

kyessenov replied:
It's a superset of attributes here https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/rbac_filter#condition. I can fill up the missing ones section once we have an overview doc in place for Wasm.

gbrail replied:
Yes -- I plan to re-use the code used in WASM support. I'm not sure where that's documented or how to add a link here that would point to the right place, however. Any ideas are appreciated!

mattklein123 replied:
WASM has little/no documentation currently which is a known deficiency, so it would be great if you all could work together to get this all documented in some sane way.

htuch commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+// [#next-free-field: 9]
+message ProcessingMode {
+  enum SyncMode {
+    // The default SyncMode depends on which part of the message is being set.
```
I think just call it DEFAULT or DEFAULT_MODE then. It's too easy to confuse with SYNC below. Same with BODY_DEFAULT.

gbrail replied:
Agreed.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+
+// Change the processing mode of the filter. If "duration" is set,
+// then the change is a temporary change that will be un-done after the specified
+// timeout. (If multiple requests arrive after a timer has been set, then the results
```
This only applies to the current stream, right?

gbrail replied:
No, it's either permanent (for the Envoy server that received the message) or subject to a duration. This message is only used with the "mode_change" field in the ProcessingResponse.

There is also a "mode_override" field in the ProcessingResponse that changes the mode only for this stream.

mattklein123 replied:
Per above this seems super complicated. Do we need this in the initial impl? I can also foresee all kinds of race conditions if we can go from async back to sync for example.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+}
+
+message BodyResponse {
+  // Add or replace HTTP headers.
```
So, are we implicitly relying on the gRPC stream flow events (e.g. waiting for remote server response) to guide flow control (e.g. StopIteration, Continue) on the Envoy side? This makes sense, but I wonder if there is finer grained control over stop iteration, e.g. buffering, watermarking that we might need. In general I'd recommend that @alyssawilk takes a pass on this for watermark implications.

alyssawilk replied:
I would assume we're doing the same thing we're doing for all async client operations here, which is relying on gRPC/H2 flow control for send and receive.  Overflow semantics seem documented on the request path, and I assume if any mutations on the gRPC resposne result in overflow we'd have the same close logic but it seems worth commenting.

gbrail replied:
In general, I expect that, if the processing mode is "SYNC," then Envoy would send a message on the stream and suspend filter iteration until it gets back a response, so yes, that's the flow control mechanism. If the processing mode is ASYNC, however, then we want to avoid blowing up all the buffers in Envoy. I saw recently that each gRPC stream has a watermark on it, so perhaps that's a good way to go.

I will certainly need advice on the specific values to use, and those might best be done with a code review by @alyssawilk when we get to some implementation.

htuch reviewed (commented):
Looks great, flushing a few comments.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,89 @@
+  // Specifies which parts of the HTTP request and response are sent to the remote
+  // server by default, whether they are sent synchronously or asynchronously,
+  // and whether the body is sent in chunks or buffers. The default processing
+  // mode sends request and response headers synchronously and nothing else.
```
I find the phrasing on this one confusing.  I'd read synchronous send as blocking Envoy until an RPC is received rather than blocking the stream.
I think we can just say "gives default options for how headers and body are sent, see ProcessingMode for details"

gbrail replied:
I'm all for making the doc more readable like this.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+
+// [#next-free-field: 9]
+message ProcessingMode {
+  enum SyncMode {
```
If this affects more than headers, I think the comments should be updated.

gbrail replied:
Ack.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+    // The default SyncMode depends on which part of the message is being set.
+    BODY_DEFAULT = 0;
+
+    // Send each chunk of the body separately to the server.
```
I'd be inclined to call this streaming.  Chunked could be confused with HTTP/1.1 chunks and we're not actually sending a chunk at a time.

gbrail replied:
Makes sense -- let's not use the "chunk" term.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+    CHUNKED = 1;
+
+    // Buffer the message body in memory and send the entire body in one
+    // chunk. If the body exceeds the configured buffer limit, then the
```
chunk -> rpc?

gbrail replied:
Ack.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+  BodyMode request_body_chunk_mode = 4;
+
+  // How do handle the response body. Default is "SKIP".
+  ProcessingMode response_body_mode = 5;
```
body_send_move vs body_chunk_mode for body stuff?

gbrail replied:
Yes -- and I renamed BodyMode to BodySendMode as well for clarity.

alyssawilk commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  // How to handle the request header. Default is "SYNC".
+  ProcessingMode request_header_mode = 1;
```
Is it legal to have async headers and sync body?  Mixing and matching seems confusing and dangerous.

Alternately, would it make sense to have  one blocking/non-blocking boolean for request/response  and "send or don't send" booleans for different parts of the request/response?

ahedberg replied:
+1, mixing and matching sync vs async on the same service feels confusing and error-prone.

Are we sure we even need this flexibility from the get-go? i.e. can the initial implementation be "always wait for a response before continuing with the filter chain" and we can add the fire-and-forget functionality later? I'm having trouble imagining a use case for sending the headers to an external service but not caring what the response is anyway.

alyssawilk commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+// message.
+message HeaderResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
```
Might be worth mentioning if the peer is responsible for handling appending by taking incoming headers and adding their own.   Alternately should there be explicit
append
add_or_replace
remove
options?

gbrail replied:
HeaderValueOption has an "append" flag so you can choose whether to append or add/replace using this protocol already. I like re-using an existing type here instead of having separate fields.

alyssawilk commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
```
should the BodyResponse contain a HeaderResponse message rather than duplicate fields?

gbrail replied:
Reducing duplication sounds good.

In practice, we can't modify the headers once they're sent further down the filter chain. I know that this can work when the entire body is buffered, and I suppose it can work the first time the a body filter callback is called, but after that if the user replies with a request to modify headers, we will have to ignore it. Not sure how to document all that.

alyssawilk commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+  repeated string remove_headers = 2;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 3;
```
How does this work with streaming?  
If Envoy sends chunk A B C as they arrive, and it gets back a response I assume the modification is for chunk A not chunk C (which is the last to be delivered)

Looking at the definition it looks more like a general offset thing than anything else.

gbrail replied:
It's right now a request / response protocol. If the sync mode is "SYNC," then Envoy sends a HttpBody message each time a "Body" is called on the filter and suspends processing until it gets a BodyResponse message. So, the BodyResponse always modifies the last body that was received by the filter, and there's no pipelining. (Although I suppose that we could add a sequence number fo each HttpBody message if you think pipelining is important, or allow pipelining in general.)

So there's one BodyMutation per HttpBody.

I could see the body_mutation field being repeated, however, but I'm not sure exactly what the use case would be.

alyssawilk commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+
+// Change all or part of the message body by replacing the portion
+// of the last chunk received denoted by "offset" and "length" with "body."
+message BodyMutation {
```
can we think offhand of any case where we want to modify a few bytes?  I'm used to most filters passing through the body, and a very few doing transformation (usually compression or decompression).  If there are delta-use cases please ignore this comment but I wonder if we're just better off saying the server can either have no effect on body or send its own response body.

gbrail replied:
I think that replacing the whole body, in most cases as one single buffer, will be pretty common in the kinds of use cases I'm used to from the API management and integration world. For instance, transform a message body from one format to another (JSON to XML to protobuf and so on), or run something like XSLT or JavaScript on the message body for a larger transformation.

So yes, I think that replacing the whole body will be pretty common.

Modifying a few bytes is going to be pretty uncommon, I agree, and I could see the need to just replace the whole body, which would eliminate a set of errors.

This bit is supposed to work like the WASM ABI, which lets a filter append or prepend or modify the middle. I have mixed feelings about whether or not we need that complexity here.

alyssawilk commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+
+// Change all or part of the message body by replacing the portion
+// of the last chunk received denoted by "offset" and "length" with "body."
+message BodyMutation {
```
Also I assume we could remove a body if there was one present by just saying offset 0, length = numeric_limits::max<something>, and no bytes, but how about adding a body to a response where there previously was none?  We have Envoy filters which do this.  Ditto adding trailers.

gbrail replied:
Yes. I don't think that this protocol addresses that and we need to figure out a way. I'm not sure how right now!

alyssawilk replied:
@PiotrSikora does this exist for WASM?

alyssawilk reviewed (commented):
This is super cool.  I don't have many thoughts w.r.t. flow control, but I have some other thoughts w.r.t. HTTP APIs.

I know he's already been tagged but re-tagging @PiotrSikora since I think this could be used for running WASM sandboxed out of process and we want to make sure it plays well.
Also @ahedberg to review or delegate since I think this may end up being the API her project uses in the distant future.

gbrail commented:
On the implementation plan -- my strategy with this PR was to spec out the
whole protocol at the start. I am very open to not implementing it all at
first.

In that case, is it better to submit a fully-formed .proto and not
implement it all, or to gradually build out the .proto during
implementation? (I also assume that it will take many PRs to get to
anything at all and that this is what the "work in progress" flags on the
filter are for.)

On the design -- there's a doc with more details -- it probably got buried
in this long comment stream:

https://docs.google.com/document/d/1IZqm5IUnG9gc2VqwGaN5C2TZAD9_QbsY9Vvy5vr9Zmw/edit#heading=h.gfz6mi5p212e

Early on we had some comments from a team who wanted the ability to monitor
the flow of traffic for anomalies or threats, but then if a specific threat
is detected, switch to synchronous mode temporarily in order to actually
block the threat. We also had some discussion as to whether the best way to
do this would be via this filter or by pushing an Envoy configuration
change to enable it.

I also can think of a number of use cases in which a filter will want the
"ProcessingMode" outside the sync/async requirement to change per message
-- for instance, a filter that represents some fairly complex mediation
system might want to read and transform the bodies of some HTTP requests
but not others depending on the URL or other headers.

What I can't think of is a filter that wants to control which parts of the
message it sees, and wants to select "async" or "sync" for different parts
-- it may make more sense to have a single boolean for the whole service
that denotes async versus sync mode.

Sorry for the long answer!

On Wed, Oct 28, 2020 at 8:33 AM ahedberg <notifications@github.com> wrote:

> *@ahedberg* commented on this pull request.
> ------------------------------
>
> In
> api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r513541710>:
>
> > +    // Send each chunk of the body separately to the server.
> +    CHUNKED = 1;
> +
> +    // Buffer the message body in memory and send the entire body in one
> +    // chunk. If the body exceeds the configured buffer limit, then the
> +    // downstream system will receive an error.
> +    BUFFERED = 2;
> +
> +    // Buffer the message body in memory and send the entire body in one
> +    // chunk. If the body exceeds the configured buffer limit, then the body contents
> +    // up to the buffer limit will be sent.
> +    BUFFERED_PARTIAL = 3;
> +  }
> +
> +  // How to handle the request header. Default is "SYNC".
> +  ProcessingMode request_header_mode = 1;
>
> +1, mixing and matching sync vs async on the same service feels confusing
> and error-prone.
>
> Are we sure we even need this flexibility from the get-go? i.e. can the
> initial implementation be "always wait for a response before continuing
> with the filter chain" and we can add the fire-and-forget functionality
> later? I'm having trouble imagining a use case for sending the headers to
> an external service but not caring what the response is anyway.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#discussion_r513541710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I2ZZ73Y3OTD5PEWS3QLSNA2WHANCNFSM4R5FQWUA>
> .
>

mattklein123 commented:
> On the implementation plan -- my strategy with this PR was to spec out the
whole protocol at the start. I am very open to not implementing it all at
first.

I left some comments on your doc, but I think it's worth spending time in speciing out the whole thing to make sure we all agree on the shape. Then I agree we don't need or want to implement it all at once.

> In that case, is it better to submit a fully-formed .proto and not
implement it all, or to gradually build out the .proto during
implementation? (I also assume that it will take many PRs to get to
anything at all and that this is what the "work in progress" flags on the
filter are for.)

I could go either way on this. My preference for this would be to agree on the roughly final proto that we want to build towards, and then either commit it with "not implemented hide" tags as things get implemented, or just leave this PR open as a guide and do smaller PRs as we go along.

On the topic of async vs. sync, I agree that we almost definitely will want this for analytics, security auditing, etc. Whether we need to implement it right away is a different question but we should make sure that the API shape supports it. Now, I do think we need to reconcile a fully async flow in this filter with the tap filter, as it starts to look pretty similar, and maybe as we move most of the tap filter matching into the core, the existing tap filter gets deprecated in favor of this one which is a more general solution. (I commented on the likely need for matching in this filter in the doc.)

mattklein123 commented:
@gbrail has this been updated since I looked and is ready for another pass? In general a friendly request to avoid force pushing at this point so it's easier to view deltas. Thank you.

gbrail commented:
I have another proto update I'd like to push to address more comments,
which I'll do shortly.

On Wed, Oct 28, 2020 at 1:27 PM Matt Klein <notifications@github.com> wrote:

> @gbrail <https://github.com/gbrail> has this been updated since I looked
> and is ready for another pass? In general a friendly request to avoid force
> pushing at this point so it's easier to view deltas. Thank you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#issuecomment-718188957>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I26OWJUTRYVWFVT3K33SNB5DTANCNFSM4R5FQWUA>
> .
>

mattklein123 commented:
> I have another proto update I'd like to push to address more comments,
which I'll do shortly.

Sounds good please ping me and I will take another pass then!

gbrail commented:
@mattklein123 I updated the proto (I won't force-push it again) and it's ready for another look. Thanks!

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,87 @@
+  // The filter supports both the "Envoy" and "Google" gRPC clients.
+  config.core.v3.GrpcService grpc_service = 1;
+
+  // If true, then do nothing if the remote service RPC fails.
+  // Otherwise, return a 500 error if the gRPC call fails.
```
I think I commented on this in the last pass, but can this be fleshed out a bit more? There are lots of failure scenarios, including after the response headers have already been sent. I assume in that case the stream would be reset? Either way can you beef this up and clarify?

gbrail replied:
I'm updating the comments.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,87 @@
+// :ref:`ProcessingResponse <envoy_v3_api_msg_service.ext_proc.v3alpha.ProcessingResponse>`.
+
+// [#next-free-field: 7]
+message ExternalProcessor {
```
A few more things I would add in the filter config mesage:
1) A plan for supporting per-route config per the design doc. In some filters this is the same message. In other filters sometimes it makes more sense to have a different message for per-route config and a common config message shared between them.
2) A runtime kill switch field that allows the processor to be turned off and on via runtime. 

See other filter configs for examples of this kind of thing.

gbrail replied:
I created a per route config message and copied a number of capabilities from ext_auth, including the ability to send only a fraction of the traffic and also to disable the filter entirely.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,90 @@
+  // Control how headers and trailers are handled
+  enum HeaderSendMode {
+    // The default HeaderSendMode depends on which part of the message is being
+    // processed. By default, request and response headers are sent, while
```
```suggestion
    // processed. By default, request and response headers are sent,
```

Stepping back this comment is confusing to me. Why is body referenced? I assume we are just talking about headers and trailers here and trailers are default skipped?

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  // How to handle the request header. Default is "SEND".
+  HeaderSendMode request_header_mode = 1;
```
Can you add defined only enum validations for all of these fields?

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,90 @@
+
+  // Control how the request and response bodies are handled
+  enum BodySendMode {
+    // Do not send the body at all. This is the default
+    NONE = 0;
+
+    // Stream the body to the server in pieces as they arrive at the
+    // proxy. This is the default.
```
These both say default. Which is the default?

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/processing_mode.proto:
```diff
@@ -0,0 +1,90 @@
+  // processing, or asynchronously, in which Envoy will continue with filter
+  // processing after sending the pieces.
+  enum AsyncMode {
+    // Send message parts asynchronously. This is the default.
```
I think you have async/sync swapped here?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+// to an external processing server.
+// [#next-free-field: 8]
+message ProcessingRequest {
+  // If true, then the server must either respond with exactly one ProcessingResponse
+  // message or close the stream. If false, the server must not respond with
+  // an additional message, although it may still close the stream.
+  // How this is set depends on how the ProcessingMode is configured, or if
+  // the server requests a change in the processing mode.
+  bool response_required = 1;
```
Do we need to support mode switching mid-stream? If not, can we drop this field for now. If so, can we flesh out the comment here? It's a bit confusing when/why this needs to be set versus when it is implied.

mattklein123 replied:
OK I see. WDYT about just renaming this `async_mode` then to match the other config? IMO that's more clear about what the intention is and then we can make it clear in the docs what responses are then required?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+    // HeaderResponse message or close the stream.
+    HttpHeaders response_headers = 3;
+
+    // A chunk of the HTTP request body. This is only sent if the server
+    // requested it. If "response_required" is set, the server must send back
```
"only sent if the server requested it" - nit, really all of these including headers above are configured by the processing config? Should we drop this verbiage or make it clear it's set by processing mode?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
```
Per my other comment, I can buy that the server wants to change what is sent based on what it received, but do we need mode switching? This seems pretty complicated and it's unclear to me if it's really necessary?

gbrail replied:
I'd be thrilled to get rid of the whole async vs sync mode thing for now, and also get rid of the idea of a permanent "mode change" driven by the server. We can bring in async mode in the future when we have made progress on the other stuff if it's turning out to be helpful.

Let me take a pass at simplifying this tomorrow and address the other comments too.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
```
What does context mean? Do you mean stream? Also, don't you need to specify which direction for a bi-di stream?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+// This message must be sent in response to a request_headers or response_headers
+// message.
+message HeadersResponse {
+  // Instructions on how to manipulate the request headers
```
```suggestion
  // Instructions on how to manipulate the headers
```

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+message HeadersResponse {
```
I think this message you also want:
1) To be able to specify trailers
2) To be able to specify end_stream

This would effectively allow you to create a complete request/response given just the headers if you want and override everything else that might be coming.

mattklein123 replied:
Yes ^ is exactly right. It's something that can be done by existing C++ filters and is something that ends up being done occasionally. Essentially, what I'm talking about is transformations in which we do things like:
1) Make a header only request into a not header only request, and optionally add data and trailers at the same time.
2) Make a not header only request into a header only request.

The other major use case that you need to think through is something like the following:
1) ext proc server gets header only request.
2) respond and make not header only request, but don't supply enough info (body, trailers, etc.) to complete the request.
3) ext proc responds and streams body/trailer data as it becomes available.

These types of things are all things that can be done with C++ filters. Now, I don't think it's required that we support all of this on day 1, but I suggest making sure the design *can* support all of these things.

The TL/DR request is this: Study filter.h, all of the iteration return codes, and all of the async completion possibilities, and make sure they could be modeled by this API. I think this may involve a protocol that is not always request/response. It may be a request followed by a series of streamed responses for example.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,273 @@
+  HeaderMutation header_mutation = 1;
+}
+
+message BodyResponse {
```
Same comment about trailers and end_stream

mattklein123 replied:
See above, essentially we are talking about the ability to end a stream early, or *not* end a stream as the case may be.

mattklein123 reviewed (changes requested):
Thanks overall this is looking good. Fleshing out some more comments from this round. Thanks for working on this!

/wait

gbrail commented:
@mattklein123 I addressed code review comments in the latest two commits, although I still have a question about your comments regarding the HeaderResponse message. Let me know when you have time to take a look. Thanks!

gbrail commented:
Also, in general -- I feel like we have two choices -- check in the complete configuration and implement features gradually, or check in a much smaller configuration that is also implemented. I'd appreciate guidance so that I know what to work on next.

In other words, should we plan to:

1) Merge this PR, which contains the complete configuration, but with an implementation that ignores most of the configuration as we implement it a bit at a time. (This PR has a do-nothing implementation.)
2) Create a new PR that contains only a subset of the configuration but which implements it, and then add features and configuration gradually but at the same time.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,155 @@
+  // prematurely with an error, the filter will fail. Specifically, if the
+  // response headers have not yet been delivered, then it will return a 500
+  // error downstream. If they have been delivered, then instead the HTTP stream to the
+  // downstream client will be closed.
```
```suggestion
  // downstream client will be reset.
```

Same below.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,155 @@
+  // response message.
+  repeated string response_properties = 6;
+
+  // Specifies the timeout for each individual message sent on the stream. Whenever
+  // the proxy sends a message on the stream that requires a response, it will
+  // reset this timer, and will stop processing and return an error (subject
+  // to the processing mode) if the timer expires. Default is 200 ms.
```
nit: clarify this only applies in sync mode.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,155 @@
+  // If this field is not specified, the filter will be enabled for all requests.
+  config.core.v3.RuntimeFractionalPercent filter_enabled = 8;
+
+  // Specifies if the filter is enabled with metadata matcher.
+  // If this field is not specified, the filter will be enabled for all requests.
+  type.matcher.v3.MetadataMatcher filter_enabled_metadata = 9;
```
Unless you know you need this I would probably skip this for now, but I don't feel strongly about it.

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,155 @@
+  // If this field is not specified, the filter will be enabled for all requests.
+  type.matcher.v3.MetadataMatcher filter_enabled_metadata = 9;
+
+  // Specifies if the peer certificate is sent to the external service.
+  //
+  // When this field is true, Envoy will include the peer X.509 certificate, if available, in the
+  // :ref:`certificate<envoy_api_field_service.auth.v3.AttributeContext.Peer.certificate>`,
+  // which is part of the "client_info" field.
+  bool include_peer_certificate = 10;
```
I feel like this would be covered by request properties?

mattklein123 commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,155 @@
+}
+
+// Overrides that may be set on a per-route basis
+message ExtProcOverrides {
```
I feel like this message can just be inlined above, but up to you.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+
+// For every ProcessingRequest received by the server with the "response_required" field
+// set, the server must send back exactly one ProcessingResponse message.
+// In addition, the server may send back a message with the "async_mode_change"
+// message at any time.
```
I think this is gone?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
```
I'm wondering can all of this just be folded into properties?

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,254 @@
+  BodyMutation body_mutation = 2;
+}
+
+// This message completely replaces the response.
+// It will immediately return the data specified here to the downstream
+// system and stop further filter processing.
```
I would clarify what happens if the response has already been started when this is received.

mattklein123 reviewed (changes requested):
Thanks this is getting really close. I left some more comments, but now that @htuch is back he should definitely look again.

One other comment: we are reaching the limits of GH's capability to deal with a review like this and I'm seeing a bunch of timeouts. We may be better off moving into a gdoc to lock down the protos a bit further, but up to you.

/wait

mattklein123 commented:
I'm not sure if my last set of comments went out, but the PR isn't reliably
loading for me anymore. If you see this can we move the protos to a gdoc
while we sort out some of the remaining issues and then we can open a fresh
PR?

On Tue, Nov 3, 2020 at 11:03 AM Greg Brail <notifications@github.com> wrote:

> Also, in general -- I feel like we have two choices -- check in the
> complete configuration and implement features gradually, or check in a much
> smaller configuration that is also implemented. I'd appreciate guidance so
> that I know what to work on next.
>
> In other words, should we plan to:
>
>    1. Merge this PR, which contains the complete configuration, but with
>    an implementation that ignores most of the configuration as we implement it
>    a bit at a time. (This PR has a do-nothing implementation.)
>    2. Create a new PR that contains only a subset of the configuration
>    but which implements it, and then add features and configuration gradually
>    but at the same time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#issuecomment-721289396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQDL3ABC3YSPY57Y6PQUFLSOBAVZANCNFSM4R5FQWUA>
> .
>

gbrail commented:
Yes I think that we are starting to test the limits of GitHub.

I do see your comments and I'll move the protos to a doc.

On Tue, Nov 3, 2020 at 12:29 PM Matt Klein <notifications@github.com> wrote:

> I'm not sure if my last set of comments went out, but the PR isn't reliably
> loading for me anymore. If you see this can we move the protos to a gdoc
> while we sort out some of the remaining issues and then we can open a fresh
> PR?
>
> On Tue, Nov 3, 2020 at 11:03 AM Greg Brail <notifications@github.com>
> wrote:
>
> > Also, in general -- I feel like we have two choices -- check in the
> > complete configuration and implement features gradually, or check in a
> much
> > smaller configuration that is also implemented. I'd appreciate guidance
> so
> > that I know what to work on next.
> >
> > In other words, should we plan to:
> >
> > 1. Merge this PR, which contains the complete configuration, but with
> > an implementation that ignores most of the configuration as we implement
> it
> > a bit at a time. (This PR has a do-nothing implementation.)
> > 2. Create a new PR that contains only a subset of the configuration
> > but which implements it, and then add features and configuration
> gradually
> > but at the same time.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/envoyproxy/envoy/pull/13304#issuecomment-721289396>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABQDL3ABC3YSPY57Y6PQUFLSOBAVZANCNFSM4R5FQWUA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/13304#issuecomment-721358252>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I2YKUOXRW2MPUW7MCJTSOBR2TANCNFSM4R5FQWUA>
> .
>

gbrail commented:
165 comments later we seem to have broken GitHub. A cleaned up PR has been recreated here:

https://github.com/envoyproxy/envoy/pull/13893
